[SYCL][DOC] Introduce property to switch off error on cycle (#128)
reble/llvm: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

};

namespace property {

namespace graph {

> class no_cycle_error_check {

Reviewer: I think the word error in the property name is a bit superfluous and `no_cyle_check` would be a bit more concise.

> class no_cycle_error_check {

Author: ```suggestion
class no_cycle_check {
```

  public:

>     no_cycle_error_check() = default; 

Author: ```suggestion
    no_cycle_check() = default; 
```

* Throws synchronously with error code `invalid` if `src` and `dest`
  are the same node.
  
* Throws synchronously with error code `invalid` if the resulting dependency would

>   lead to a cycle.

Author: ```suggestion
  lead to a cycle. This error is omitted when `property::graph::no_cycle_check` is set.
```

  the same order.
|===

> ==== Graph Properties

Reviewer 2: The description of the `propList` parameter to the constructor now needs updated (line 507), it currently says
```
* `propList` - Optional parameter for passing properties. No `command_graph`
   constructor properties are defined by this extension.
```
Rather than saying "No `command_graph` constructor properties are defined by this extension.", i'd mention to refer to this section.

Author: Good catch, updated the description here.

> There are the following properties defined that can be passed to a `command_graph`

Author: ```suggestion
There is the one following property defined that can be passed to a `command_graph`
```

on construction via the property list parameter.

> 1. `property::graph::no_cycle_error_check` - This property disables any checks if

Author: ```suggestion
1. `property::graph::no_cycle_check` - This property disables any checks if
```

   a newly added dependency will lead to a cycle in a specific `command_graph`.
   As a result, no errors are reported when a function tries to create a cyclic
   dependency. Thus, it's the user's responsibility to create an acyclic graph 
   for execution when this property is set. Executing any graph with cyclic

>    dependencies is undefined behavior.

Reviewer: I think the undefined behaviour here could be even more broad than just execution, I think creating any modifiable graph with cycles would potentially lead to undefined behaviour and break what we're doing in `finalize` etc.

>    dependencies is undefined behavior.

Author: ```suggestion
   for execution when this property is set. Creating a `command_graph` in
   executable state through `finalize` from a graph with cyclic dependencies
   is undefined behavior.
```

  the same order.
|===

==== Graph Properties

There is the one following property defined that can be passed to a `command_graph`
on construction via the property list parameter.

> 1. `property::graph::no_cycle_check` - This property disables any checks if

Reviewer 2: I think we'll need a new error listed in the Exceptions description of `make_edge()` to say that it will throw an error if adding an edge creates a cycle. This is the error checking behaviour requested, that using this property will disable.

   a newly added dependency will lead to a cycle in a specific `command_graph`.
   As a result, no errors are reported when a function tries to create a cyclic
   dependency. Thus, it's the user's responsibility to create an acyclic graph 
   for execution when this property is set. Creating a `command_graph` in
   executable state through `finalize` from a graph with cyclic dependencies

>    is undefined behavior.

Author: ```suggestion
   is not allowed and results in undefined behavior.
```
